numba/compiler_machinery.py

```diff
@@ -0,0 +1,376 @@
+from __future__ import print_function, absolute_import
+
+import timeit
+from abc import abstractmethod, ABCMeta
+from collections import namedtuple, OrderedDict
+import inspect
+import traceback
+from numba.compiler_lock import global_compiler_lock
+from numba import errors
+from . import config, utils, transforms, six
+from .tracing import event
+from .postproc import PostProcessor
+
+# terminal color markup
+_termcolor = errors.termcolor()
+
+
+class SimpleTimer():
+    """
+    A simple context managed timer
+    """
+
+    def __enter__(self):
+        self.ts = timeit.default_timer()
+        return self
+
+    def __exit__(self, *exc):
+        self.elapsed = timeit.default_timer() - self.ts
+
+
+@six.add_metaclass(ABCMeta)
+class CompilerPass(object):
+
+    @abstractmethod
+    def __init__(self, *args, **kwargs):
```

Comment:
Why does `__init__` has to be an `abstractmethod`? A lot of times it's just a passthru to the parent's implementation.

Reply:
I don't think it needs to be. I tried a bunch of different ways of describing this class and the pipeline and I think that this is an artifact from when pass registration required an instance of a pass to be registered. I was thinking about how pass classes might be extended, and was considering the merits of inheritance over composition and settled for the latter as most of the time passes are pretty unique should be state free by design.